kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionCoordinator.java

```diff
@@ -128,12 +129,31 @@ public static TransactionCoordinator of(String tenant,
         String namespacePrefixForUserTopics = MetadataUtils.constructUserTopicsNamespace(tenant, kafkaConfig);
         TransactionStateManager transactionStateManager =
                 new TransactionStateManager(transactionConfig, txnTopicClient, scheduler, time);
+        final Supplier<ProducerIdManager> producerIdManagerSupplier = () -> {
+            if (kafkaConfig.isKafkaTransactionProducerIdsStoredOnPulsar()) {
+                return new PulsarStorageProducerIdManagerImpl(
+                        transactionConfig.getTransactionProducerIdTopicName(), txnTopicClient.getPulsarClient());
+            } else {
+                return new ProducerIdManagerImpl(transactionConfig.getBrokerId(), metadataStore);
+            }
+        };
         ProducerIdManager producerIdManager;
-        if (kafkaConfig.isKafkaTransactionProducerIdsStoredOnPulsar()) {
-            producerIdManager = new PulsarStorageProducerIdManagerImpl(
-                    transactionConfig.getTransactionProducerIdTopicName(), txnTopicClient.getPulsarClient());
+        final String producerIdManagerClassName = kafkaConfig.getKopProducerIdManagerClassName();
+        if (producerIdManagerClassName == null) {
+            producerIdManager = producerIdManagerSupplier.get();
         } else {
-            producerIdManager = new ProducerIdManagerImpl(transactionConfig.getBrokerId(), metadataStore);
+            try {
+                producerIdManager = (ProducerIdManager)
+                        Class.forName(producerIdManagerClassName).getDeclaredConstructor().newInstance();
```

Comment:
I think that at least we have to pass the KafkaServiceConfiguration object,

the problem is that we need access to some env object, like the PulsarClient

Reply:
I think we need to modify the `ProducerIdManager` interface as well. Ideally I think the `initialize` method should be:

```java
    // If we use `KafkaServiceConfiguration`, maybe we'd better put it in another module like `kop-common`
    CompletableFuture<Void> initialize(Properties props);
```

I'm not sure if it would be proper to pass a `BrokerService` object because the `PulsarClient` or `MetadataStore` object that can be retrieved from it. My initial purpose is to separate these objects from the dependency on `pulsar-broker` dependency, i.e. if you want to use ZK/Etcd/Pulsar as the pid storage, you must create the corresponding client according to your own client configs.

BTW, I think it should be carefully considered and it's not an urgent task so I mark it as drafted and might continue in future.